codecs/zarrs.zfp/README.md

```diff
@@ -0,0 +1,88 @@
+# ZFP codec
+
+Defines a `array -> bytes` codec that compresses chunks using the [zfp](https://github.com/LLNL/zfp) algorithm.
+
+## Codec name
+
+The value of the `name` member in the codec object MUST be `zarrs.zfp`.
+
+## Configuration parameters
+
+The configuration of this codec matches the compression modes defined at <https://zfp.readthedocs.io/en/latest/modes.html#expert-mode>.
+Refer to that page for usage information.
+
+The codec has one parameter which is always required:
+- `mode` (string).
+
+The other required parameters are dependent on the `mode`:
+- `"mode": "reversible"`
+- `"mode": "expert"`
+    - `minbits` (unsigned integer)
+    - `maxbits` (unsigned integer)
+    - `maxprec` (unsigned integer)
+    - `minexp` (signed integer)
+- `"mode": "fixed_accuracy"`
+    - `tolerance` (number)
+- `"mode": "fixed_rate"`
+    - `rate` (number)
+- `"fixed_precision"`
+    - `precision` (unsigned integer)
+
+## Example
+
+For example, the array metadata below specifies that the array contains zfp compressed chunks using `fixed_accuracy` mode with a tolerance of 0.05:
+
+```json
+{
+    "codecs": [{
+        "name": "zarrs.zfp",
+        "configuration": {
+            "mode": "fixed_accuracy",
+            "tolerance": 0.05
+        }
+    }],
+}
+```
+
+More examples can be viewed in the [examples](./examples/) subdirectory.
+
+## Supported Data Types
+
+- `int32`, `uint32`, `int64`, `uint64`, `float32`, `float64`
+- `int8`, `uint8`, `int16`, `uint16` (through internal promotion 32-bit)
+
+Implementations are permitted to support additional data types that could be interpreted as the above data types (e.g. `datetime64` -> `uint64`).
```

Comment:
I think datetime64 would correspond to int64 not uint64.

Reply:
You could note that lower-precision floating point types, such as float16 and the new float-precision floating point extension types I added, can be supported by first converting to float32 in the normal way.

Lower-precision integer types, e.g. {int,uint}{2,4} can also be supported but would need to be promoted in the same way that {int,uint}{8,16} are promoted:

for intN:

```c++
int32_value = (int32_t)intN_value << (31 - N)
intN_value = (intN_t)clamp(int32_value >> (31 - N), 1<<(N-1), (1<<(N-1)) - 1)
```

for uintN

```c++
int32_value = ((int32_t)uintN_value - (1<<(N-1))) << (31 - N)
uintN_value = (uintN_t)clamp((int32_value >> (31 - N)) + (1<<(N-1)), 0, (1<<N) - 1)
```